ext/random/engine_xoshiro256starstar.c

```diff
@@ -0,0 +1,249 @@
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "php.h"
+#include "php_random.h"
+
+#include "ext/spl/spl_exceptions.h"
+#include "Zend/zend_exceptions.h"
+
+static inline uint64_t splitmix64(uint64_t *seed) {
+	uint64_t r;
+
+	r = (*seed += 0x9e3779b97f4a7c15ULL);
+	r = (r ^ (r >> 30)) * 0xbf58476d1ce4e5b9ULL;
+	r = (r ^ (r >> 27)) * 0x94d049bb133111ebULL;
+	return (r ^ (r >> 31));
+}
+
+static inline uint64_t rotl(const uint64_t x, int k) {
+	return (x << k) | (x >> (64 - k));
+}
+
+static inline uint64_t generate_state(php_random_status_state_xoshiro256starstar *s)
+{
+	const uint64_t r = rotl(s->state[1] * 5, 7) * 9;
+	const uint64_t t = s->state[1] << 17;
+
+	s->state[2] ^= s->state[0];
+	s->state[3] ^= s->state[1];
+	s->state[1] ^= s->state[2];
+	s->state[0] ^= s->state[3];
+
+	s->state[2] ^= t;
+
+	s->state[3] = rotl(s->state[3], 45);
+
+	return r;
+}
+
+static inline void jump(php_random_status_state_xoshiro256starstar *state, const uint64_t *jmp)
+{
+	uint64_t s0 = 0, s1 = 0, s2 = 0, s3 = 0;
+	uint32_t i, j;
+
+	for (i = 0; i < 4; i++) {
+		for (j = 0; j < 64; j++) {
+			if (jmp[i] & 1ULL << j) {
+				s0 ^= state->state[0];
+				s1 ^= state->state[1];
+				s2 ^= state->state[2];
+				s3 ^= state->state[3];
+			}
+
+			generate_state(state);
+		}
+	}
+
+	state->state[0] = s0;
+	state->state[1] = s1;
+	state->state[2] = s2;
+	state->state[3] = s3;
+}
+
+static inline void seed256(php_random_status *status, uint64_t s0, uint64_t s1, uint64_t s2, uint64_t s3)
+{
+	php_random_status_state_xoshiro256starstar *s = status->state;
+
+	s->state[0] = s0;
+	s->state[1] = s1;
+	s->state[2] = s2;
+	s->state[3] = s3;
+}
+
+static void seed(php_random_status *status, uint64_t seed)
+{
+	uint64_t s[4];
+	
+	s[0] = splitmix64(&seed);
+	s[1] = splitmix64(&seed);
+	s[2] = splitmix64(&seed);
+	s[3] = splitmix64(&seed);
+
+	seed256(status, s[0], s[1], s[2], s[3]);
+}
+
+static uint64_t generate(php_random_status *status)
+{
+	return generate_state(status->state);
+}
+
+static zend_long range(php_random_status *status, zend_long min, zend_long max)
+{
+	return php_random_range(&php_random_algo_xoshiro256starstar, status, min, max);
+}
+
+static bool serialize(php_random_status *status, HashTable *data)
+{
+	php_random_status_state_xoshiro256starstar *s = status->state;
+	zval z;
+	uint32_t i;
+
+	for (i = 0; i < 4; i++) {
+		ZVAL_STR(&z, php_random_bin2hex((unsigned char *) &s->state[i], sizeof(uint64_t)));
+		zend_hash_next_index_insert(data, &z);
+	}
+
+	return true;
+}
+
+static bool unserialize(php_random_status *status, HashTable *data)
+{
+	php_random_status_state_xoshiro256starstar *s = status->state;
+	zval *z;
+	uint32_t i;
+
+	for (i = 0; i < 4; i++) {
+		z = zend_hash_index_find(data, i);
+		if (!z || Z_TYPE_P(z) != IS_STRING) {
+			return FAILURE;
+		}
+
+		php_random_hex2bin(Z_STR_P(z), &s->state[i]);
+	}
+
+	return true;
+}
```

Comment:
I think you should validate the input. Here's an example serialization that is pretty broken:

```
O:32:"Random\Engine\Xoshiro256StarStar":2:{i:0;a:0:{}i:1;a:3:{i:0;s:8:"577a315d";i:1;s:16:"b87730b4708aXXXX";i:3;s:16:"b7f1b3ba29e0dc31";}}
```

I suggest the following checks:

- There are exactly 4 elements in the given hash table.
- These have the keys 0, 1, 2, 3.
- Each is a 16 byte string.
- Each character is in the range `0-9a-f`.

Adapt this as necessary for the other engines (e.g. PcgOneseq128XslRr64 only has 2 elements with keys 0, 1).

Reply:
I made a mistake in the value to be returned in case of failure. With this fix, unserialize will fail correctly.

e5ff0510842c35e04a554e3e506423d8fed04cca
ede881498ab4e57d880b3093016a288edfbab0df (more strict)